docs/proposals/31-inplace-node-update.md

```diff
@@ -0,0 +1,388 @@
+status: implementable
+authors:
+  - "@acumino"
+  - "@ary1992"
+  - "@shafeeqes"
+---
+
+# GEP-31: In-Place Node Updates of Shoot Clusters
+
+## Table of Contents
+
+- [GEP-31: In-Place Node Updates of Shoot Clusters](#gep-31-in-place-node-updates-of-shoot-clusters)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [Approach](#approach)
+    - [Prerequisites](#prerequisites)
+    - [Update Strategies](#update-strategies)
+      - [InplaceUpdate strategy](#inplaceupdate-strategy)
+      - [InplaceUpdateOnLabel strategy](#inplaceupdateonlabel-strategy)
+    - [Gardener](#gardener)
+      - [`CloudProfile` API](#cloudprofile-api)
+      - [`Shoot` API](#shoot-api)
+      - [`Worker` API](#worker-api)
+      - [`OperatingSystemConfig` API](#operatingsystemconfig-api)
+      - [Gardener Node Agent](#gardener-node-agent)
+    - [Machine Controller Manager](#machine-controller-manager)
+      - [`MachineDeployment` API](#machinedeployment-api)
+    - [Dependency Watchdog](#dependency-watchdog)
+    - [Extensions](#extensions)
+      - [OS extensions](#os-extensions)
+    - [Failures and Recovery Strategy](#failures-and-recovery-strategy)
+      - [Types of Failures During Update](#types-of-failures-during-update)
+        - [Drain Timeout/Failure](#drain-timeoutfailure)
+        - [Kubernetes Update Failures](#kubernetes-update-failures)
+        - [OS Update Failures](#os-update-failures)
+      - [Roles in Failure Handling](#roles-in-failure-handling)
+    - [Additional Validations](#additional-validations)
+  - [Future Work](#future-work)
+  - [Alternatives](#alternatives)
+    - [Using Only Gardener Node Agent](#using-only-gardener-node-agent)
+    - [Using the same name for Machine Class](#using-the-same-name-for-machine-class)
+
+## Summary
+
+Gardener should support new update strategies for nodes that allow updates to Kubernetes minor versions and/or Machine image/OS versions without requiring the deletion and recreation of the nodes. These strategies aim to minimize the overhead traditionally associated with node replacement and offer an alternative approach to updates, particularly important for physical machines or bare-metal nodes.
+
+The in-place update strategy is not intended to replace the current rolling update strategy provided by the [`machine-controller-manager`](https://github.com/gardener/machine-controller-manager). Instead, it is designed for specific scenarios where physical machines or bare metal nodes, whether manually joined or programmatically provisioned, cannot be easily rolled. Additionally, this update strategy can be useful in scenarios where rolling machines are undesirable, such as when dealing with large machines or limited resources on the cloud provider. For all other scenarios where preserving nodes is not critical, the rolling update strategy remains the preferred approach.
```

Comment:
How about `RecreateUpdate` instead of `TerminateAndRecreateUpdate`, Recreate itself implies the machine is deleted and created again. So it would be `RecreateUpdateRolling`, `RecreateUpdateOnDelete`, `RecreateUpdateOnLabel`, `InPlaceUpdateRolling`, `InPlaceUpdateOnLabel`.

Reply:
After some discussion, we agreed on allowing users to choose between the old replace and the new in-place strategy (in case there are good reasons we do not foresee today - and for backward compatibility resp. giving the users the means to go back to what we had before in case there are issues).

We also agreed on the following names:
- `ManualReplace` <-- what HANA and others asked for in the past
- `AutoReplace` <-- what we have today
- `ManualInPlace` <-- what we are planning to do
- `AutoInPlace` <-- what we are planning to do

It will be allowed to freely transition between `Manual` and `Auto` and vice versa, but...

it will **not** be allowed to transition between `Replace` and `InPlace`, because that would complicate the implementations (hash problem), but it can be added later (`InPlace` to `Replace` is easy, but so is the mitigation for the end user by creating a new worker pool; `Replace` to `InPlace` is more difficult and may require to at least once recreate the machines and only then further updates can be `InPlace`).